ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/metrics/DelayMetric.scala

```diff
@@ -0,0 +1,122 @@
+  }
+
+  private def updatedObjectives(
+      newValue: DelayMetric.Value
+  ): Map[ServiceLevelObjective[DelayMetric.Value], Option[DelayMetric.Value]] = {
+    objectives
+      .map { case (objective, currentViolatingValue) =>
+        // verify if the new value violates objective's requirements
+        if (objective.isViolatedBy(newValue)) {
+          currentViolatingValue match {
+            case None =>
+              // if the new value violates objective's requirements and there is no other violating value,
+              // record the new value
+              objective -> Some(newValue)
+            case Some(currentValue) =>
+              // if the new value violates objective's requirements and there is already a value that violates
+              // requirements, record the maximum value of the two
+              objective -> Some(Ordering[Value].max(currentValue, newValue))
+          }
+        } else {
+          objective -> currentViolatingValue
+        }
+      }
+  }
+
+  override def periodicValue(): (Metric[T], DelayMetric.Value) = {
+    val value: DelayMetric.Value = DelayMetric.Value(periodicMeanDelay.map(_.getSeconds))
+    val updatedMetric = this.copy(
+      delaysInCurrentInterval = List.empty,
+      objectives = updatedObjectives(value),
+    )
+    (updatedMetric, value)
+  }
+
+  override def finalValue(totalDurationSeconds: Double): DelayMetric.Value =
+    DelayMetric.Value(None)
+
+  override def violatedObjectives
+      : Map[ServiceLevelObjective[DelayMetric.Value], DelayMetric.Value] =
+    objectives
+      .collect {
+        case (objective, value) if value.isDefined => objective -> value.get
+      }
+
+  private def periodicMeanDelay: Option[Duration] =
+    if (delaysInCurrentInterval.nonEmpty)
+      Some(
+        delaysInCurrentInterval
+          .reduceLeft(_.plus(_))
+          .dividedBy(delaysInCurrentInterval.length.toLong)
+      )
+    else None
+}
+
+object DelayMetric {
+
+  def empty[T](
+      recordTimeFunction: T => Seq[Timestamp],
+      objectives: List[ServiceLevelObjective[Value]],
+      clock: Clock,
+  ): DelayMetric[T] =
+    DelayMetric(
+      recordTimeFunction = recordTimeFunction,
+      clock = clock,
+      objectives = objectives.map(objective => objective -> None).toMap,
+    )
+
+  final case class Value(meanDelaySeconds: Option[Long]) extends MetricValue {
+    override def formatted: List[String] =
+      List(s"mean delay: ${meanDelaySeconds.getOrElse("-")} [s]")
+  }
+
+  object Value {
+    implicit val valueOrdering: Ordering[Value] = (x: Value, y: Value) => {
+      (x.meanDelaySeconds, y.meanDelaySeconds) match {
+        case (Some(xx), Some(yy)) =>
+          if (xx < yy) -1
+          else if (xx > yy) 1
+          else 0
+        case (Some(_), None) => 1
+        case (None, Some(_)) => -1
+        case (None, None) => 0
+      }
+    }
+  }
+
+  final case class MaxDelay(maxDelaySeconds: Long) extends ServiceLevelObjective[Value] {
```

Comment:
If we're at it, maybe move this guy together with `ServiceLevelObjective` or in a separate file.

Reply:
Fixed 👍